components/ScriptLibrary/include/LHVMFile.h

```diff
@@ -0,0 +1,140 @@
+/******************************************************************************
+ * Copyright (c) 2018-2024 openblack developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *******************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+
+#include <array>
+#include <filesystem>
+#include <string>
+#include <vector>
+
+#include "LHVMTypes.h"
+
+namespace openblack::LHVM
+{
+
+struct LHVMHeader
+{
+	std::array<char, 4> magic;
+	LHVMVersion version;
+};
+
+class LHVMFile
+{
+protected:
+	static constexpr const std::array<char, 4> k_Magic = {'L', 'H', 'V', 'M'};
+
+	/// True when a file has been loaded
+	bool _isLoaded {false};
+
+	std::filesystem::path _filename;
+
+	LHVMHeader _header;
+	std::vector<std::string> _variablesNames;
+	std::vector<VMInstruction> _instructions;
+	std::vector<uint32_t> _autostart;
+	std::vector<VMScript> _scripts;
+	std::vector<char> _data;
+	bool _hasStatus {false};
+	VMStack _stack;
+	std::vector<VMVar> _variableValues;
+	std::vector<VMTask> _tasks;
+	uint32_t _ticks {0};
+	uint32_t _currentLineNumber {0};
+	uint32_t _highestTaskId {0};
+	uint32_t _highestScriptId {0};
+	uint32_t _executedInstructions {0};
+
+	/// Error handling
+	void Fail(const std::string& msg);
+
+	/// Read file from the input source
+	void ReadFile(std::istream& stream);
+
+	void LoadVariablesNames(std::istream& stream, std::vector<std::string>& variables);
+	void LoadCode(std::istream& stream);
+	void LoadAuto(std::istream& stream);
+	void LoadScripts(std::istream& stream);
+	VMScript LoadScript(std::istream& stream);
+	void LoadData(std::istream& stream);
+	bool LoadStatus(std::istream& stream);
+	bool LoadStack(std::istream& stream, VMStack& stack);
+	std::vector<VMVar> LoadVariableValues(std::istream& stream);
+	void LoadTasks(std::istream& stream);
+	VMTask LoadTask(std::istream& stream);
+	void LoadRuntimeInfo(std::istream& stream);
+
+public:
+	LHVMFile()
+	    : _header {.magic = {' ', ' ', ' ', ' '}, .version = LHVMVersion::BlackAndWhite} {};
```

Comment:
What does it mean to have `    ` (four space chars) as magic? You could also just inline specify them on line 26.
All that being said. I don't really see the usefulness to store this after reading it.

Reply:
There isn't a real need to store it, but the filed is part of the `LHVMHeader` struct. We could get rid of the `LHVMHeader _header` field in the `LHVMFile` class and replace it with `LHVMVersion version` (at the moment we don't even need to store the version, but it could be useful if we will ever support future VM versions).
As you said, the magic string can be read and checked without storing, and also hardcoded in the write method (when we'll implement it).